_posts/2022-09-07-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,161 @@
+
+However, when using via-IR code generation the entire contract is first translated to Yul and then optimized as a whole. In this case, 1-3 can happen in solidity code and only the function called in (2) needs to involve an inline assembly block calling ``return(...)`` or ``stop()``. Note that due to inlining, this inline assembly block can also merely occur in another nested function call, i.e. any call in (2) that can both return to the caller and terminate via assembly is susceptible.
+
+The following is a minimal example that will exhibit the bug if compiled via-IR with enabled optimizer:
+
+```
+contract C {
+	uint public x;
+	function f(bool a) public {
+		x = 1; // This write used to be removed by the Yul optimizer.
+		g(a);
+		x = 2;
+	}
+	function g(bool a) internal {
+		// The relevant part of this function is that it can both return to the caller and terminate the transaction.
+		// The bug will show its effects in the cases in which the transaction is terminated (i.e. if a is false).
+		// In this case the write x = 1 above will be missing.
+		if (a) return;
+        assembly { return(0,0) }
+	}
+}
+```
+
+Compiling the above via-IR with enabled optimizer will result in ``f(false)`` to incorrectly terminate the transaction without modifying ``x``.
+
+For a concrete contract, the bug may still be prevented, in case the optimizer inlines the function before running the problematic optimizer step, e.g. the following contract is almost identical, but unaffected, since ``g()`` will be inlined (in the snippet above the ``return;`` prevents easy inlining):
+
+```
+contract C {
+	uint public x;
+	function f(bool a) public {
+		x = 1;
+		g(a);
+		x = 2;
+	}
+	function g(bool a) internal {
+		if (!a)
+			assembly { return(0,0) }
+	}
+}
+```
+
+To check whether your contract is actually affected, you need to trigger the condition leading to the assembly ``return(...)`` or ``stop()`` and verify that all storage writes that should have happened in the code leading to it are correctly performed.
+
+   
+## Technical Details
+
+The Yul optimizer step responsible for the bug is meant to remove storage writes that it can determine to be redundant. A storage write is considered redundant in either of the following two cases:
+- A subsequent write unconditionally overwrites the written value.
+- The transaction unconditionally reverts after the write.
+
+So similarly to above, the general pattern is:
+1. a storage write, then
+2. potentially unrelated code, and finally 
+3. either another storage write overwriting the write in (1) or an unconditional revert.
```

Comment:
(1) and (2) also have to happen within the same function or both be in the top-level block, right?

Reply:
I think this is the last important thing to discuss here. So is what I said above true? If it is, then I think it makes it significantly easier to verify if you code is affected so I think it's worth mentioning.

Reply:
Isn't that more confusing than helpful to mention? I mean, it's true that that's a prerequisite for the optimizer step - but in reality inlining will mess with this anyways...

Reply:
Ah, good point about inlining. But if not for that I think it would be helpful. I.e you don't have to consider if you're doing these things in some nested call. Only at the same level.

But yeah, I guess this really applies only to Yul and at the Solidity level it might be harder to tell if code will end up at the same level or not.

Reply:
The Yul optimizer will run the Yul inliner before it runs the redundant store eliminator... so even on the Yul level, it doesn't help that much - it'd only really be helpful if you run the redundant store eliminator in isolation or something like that, which is rather obscure :-).